server/src/main/java/org/apache/iotdb/db/qp/utils/WildcardsRemover.java

```diff
@@ -0,0 +1,160 @@
+
+    // Dataset of last query actually has only three columns, so we shouldn't limit the path num
+    // while constructing logical plan
+    // To avoid overflowing because logicalOptimize function may do maxDeduplicatedPathNum + 1, we
+    // set it to Integer.MAX_VALUE - 1
+    maxDeduplicatedPathNum =
+        queryOperator.isLastQuery()
+            ? Integer.MAX_VALUE - 1
+            : QueryResourceManager.getInstance().getMaxDeduplicatedPathNum(fetchSize);
+    soffset = queryOperator.getSeriesOffset();
+    offset = soffset;
+
+    final int slimit = queryOperator.getSeriesLimit();
+    limit = slimit == 0 || maxDeduplicatedPathNum < slimit ? maxDeduplicatedPathNum + 1 : slimit;
+
+    consumed = 0;
+  }
+
+  public WildcardsRemover(ConcatPathOptimizer concatPathOptimizer) {
+    this.concatPathOptimizer = concatPathOptimizer;
+
+    maxDeduplicatedPathNum = Integer.MAX_VALUE - 1;
+    soffset = 0;
+    limit = maxDeduplicatedPathNum + 1;
+    consumed = 0;
+  }
+
+  public List<PartialPath> removeWildcardFrom(PartialPath path) throws LogicalOptimizeException {
+    try {
+      Pair<List<PartialPath>, Integer> pair =
+          concatPathOptimizer.removeWildcard(path, limit, offset);
+
+      consumed += pair.right;
+      if (offset != 0) {
+        int delta = offset - pair.right;
+        offset = Math.max(delta, 0);
+        if (delta < 0) {
+          limit += delta;
+        }
+      } else {
+        limit -= pair.right;
+      }
+
+      return pair.left;
+    } catch (MetadataException e) {
+      throw new LogicalOptimizeException("error occurred when removing star: " + e.getMessage());
+    }
+  }
+
+  public List<List<Expression>> removeWildcardsFrom(List<Expression> expressions)
+      throws LogicalOptimizeException {
+    List<List<Expression>> extendedExpressions = new ArrayList<>();
+
+    boolean atLeastOneSeriesNotExisted = false;
+    for (Expression originExpression : expressions) {
+      List<Expression> actualExpressions = new ArrayList<>();
+      originExpression.removeWildcards(
+          new WildcardsRemover(concatPathOptimizer), actualExpressions);
+      if (actualExpressions.isEmpty()) {
+        atLeastOneSeriesNotExisted = true;
+        break;
+      }
+      extendedExpressions.add(actualExpressions);
+    }
+    if (atLeastOneSeriesNotExisted) {
+      return Collections.emptyList();
+    }
+
+    List<List<Expression>> actualExpressions = new ArrayList<>();
+    ConcatPathOptimizer.cartesianProduct(
+        extendedExpressions, actualExpressions, 0, new ArrayList<>());
+
+    List<List<Expression>> splitExpressions = new ArrayList<>();
+    for (List<Expression> actualExpression : actualExpressions) {
+      if (offset != 0) {
+        --offset;
+        continue;
+      } else if (limit != 0) {
+        --limit;
+      } else {
+        break;
+      }
+      splitExpressions.add(actualExpression);
+    }
+    consumed += actualExpressions.size();
+    return splitExpressions;
```

Comment:
From my point of view, `extendedExpressions`, `actualExpressions`, `splitExpressions`，I can't get the meaning of each name directly.
Therefore, we need add more comments or abstract a private method to do that with a clear method name.

Reply:
More comments are added. Thanks for reviewing!